{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bde3ef4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Merged Data Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 69728 entries, 0 to 69727\n",
      "Data columns (total 48 columns):\n",
      " #   Column                      Non-Null Count  Dtype  \n",
      "---  ------                      --------------  -----  \n",
      " 0   DATE ADDED TO CATALOG       69728 non-null  object \n",
      " 1   PUBMEDID                    69728 non-null  int64  \n",
      " 2   FIRST AUTHOR                69728 non-null  object \n",
      " 3   DATE                        69728 non-null  object \n",
      " 4   JOURNAL                     69728 non-null  object \n",
      " 5   LINK                        69728 non-null  object \n",
      " 6   STUDY                       69728 non-null  object \n",
      " 7   DISEASE/TRAIT               69728 non-null  object \n",
      " 8   INITIAL SAMPLE SIZE         69728 non-null  object \n",
      " 9   REPLICATION SAMPLE SIZE     69728 non-null  object \n",
      " 10  REGION                      69728 non-null  object \n",
      " 11  CHR_ID                      69728 non-null  object \n",
      " 12  CHR_POS                     69728 non-null  object \n",
      " 13  REPORTED GENE(S)            69728 non-null  object \n",
      " 14  MAPPED_GENE                 69728 non-null  object \n",
      " 15  UPSTREAM_GENE_ID            69728 non-null  object \n",
      " 16  DOWNSTREAM_GENE_ID          69728 non-null  object \n",
      " 17  SNP_GENE_IDS                69728 non-null  object \n",
      " 18  UPSTREAM_GENE_DISTANCE      69728 non-null  float64\n",
      " 19  DOWNSTREAM_GENE_DISTANCE    69728 non-null  float64\n",
      " 20  STRONGEST SNP-RISK ALLELE   69728 non-null  object \n",
      " 21  SNPS                        69728 non-null  object \n",
      " 22  MERGED                      69728 non-null  int64  \n",
      " 23  SNP_ID_CURRENT              69728 non-null  float64\n",
      " 24  CONTEXT                     69728 non-null  object \n",
      " 25  INTERGENIC                  69728 non-null  float64\n",
      " 26  RISK ALLELE FREQUENCY       69728 non-null  object \n",
      " 27  P-VALUE                     69728 non-null  float64\n",
      " 28  PVALUE_MLOG                 69728 non-null  float64\n",
      " 29  P-VALUE (TEXT)              69728 non-null  object \n",
      " 30  OR or BETA                  69728 non-null  float64\n",
      " 31  95% CI (TEXT)               69728 non-null  object \n",
      " 32  PLATFORM [SNPS PASSING QC]  69728 non-null  object \n",
      " 33  CNV                         69728 non-null  object \n",
      " 34  MAPPED_TRAIT                69728 non-null  object \n",
      " 35  MAPPED_TRAIT_URI            69728 non-null  object \n",
      " 36  STUDY ACCESSION             69728 non-null  object \n",
      " 37  GENOTYPING TECHNOLOGY       69728 non-null  object \n",
      " 38  Gene_Key                    69728 non-null  object \n",
      " 39  Entry                       62884 non-null  object \n",
      " 40  Reviewed                    62884 non-null  object \n",
      " 41  Entry Name                  62884 non-null  object \n",
      " 42  Protein names               62884 non-null  object \n",
      " 43  Gene Names                  62884 non-null  object \n",
      " 44  Organism                    62884 non-null  object \n",
      " 45  Length                      62884 non-null  float64\n",
      " 46  Involvement in disease      62884 non-null  object \n",
      " 47  Sequence_y                  62884 non-null  object \n",
      "dtypes: float64(8), int64(2), object(38)\n",
      "memory usage: 25.5+ MB\n",
      "None\n",
      "\n",
      "Duplicated columns:\n",
      "Index([], dtype='object')\n",
      "\n",
      "Missing values in each column:\n",
      "DATE ADDED TO CATALOG            0\n",
      "PUBMEDID                         0\n",
      "FIRST AUTHOR                     0\n",
      "DATE                             0\n",
      "JOURNAL                          0\n",
      "LINK                             0\n",
      "STUDY                            0\n",
      "DISEASE/TRAIT                    0\n",
      "INITIAL SAMPLE SIZE              0\n",
      "REPLICATION SAMPLE SIZE          0\n",
      "REGION                           0\n",
      "CHR_ID                           0\n",
      "CHR_POS                          0\n",
      "REPORTED GENE(S)                 0\n",
      "MAPPED_GENE                      0\n",
      "UPSTREAM_GENE_ID                 0\n",
      "DOWNSTREAM_GENE_ID               0\n",
      "SNP_GENE_IDS                     0\n",
      "UPSTREAM_GENE_DISTANCE           0\n",
      "DOWNSTREAM_GENE_DISTANCE         0\n",
      "STRONGEST SNP-RISK ALLELE        0\n",
      "SNPS                             0\n",
      "MERGED                           0\n",
      "SNP_ID_CURRENT                   0\n",
      "CONTEXT                          0\n",
      "INTERGENIC                       0\n",
      "RISK ALLELE FREQUENCY            0\n",
      "P-VALUE                          0\n",
      "PVALUE_MLOG                      0\n",
      "P-VALUE (TEXT)                   0\n",
      "OR or BETA                       0\n",
      "95% CI (TEXT)                    0\n",
      "PLATFORM [SNPS PASSING QC]       0\n",
      "CNV                              0\n",
      "MAPPED_TRAIT                     0\n",
      "MAPPED_TRAIT_URI                 0\n",
      "STUDY ACCESSION                  0\n",
      "GENOTYPING TECHNOLOGY            0\n",
      "Gene_Key                         0\n",
      "Entry                         6844\n",
      "Reviewed                      6844\n",
      "Entry Name                    6844\n",
      "Protein names                 6844\n",
      "Gene Names                    6844\n",
      "Organism                      6844\n",
      "Length                        6844\n",
      "Involvement in disease        6844\n",
      "Sequence_y                    6844\n",
      "dtype: int64\n",
      "\n",
      "Missing values after filling:\n",
      "DATE ADDED TO CATALOG         0\n",
      "PUBMEDID                      0\n",
      "FIRST AUTHOR                  0\n",
      "DATE                          0\n",
      "JOURNAL                       0\n",
      "LINK                          0\n",
      "STUDY                         0\n",
      "DISEASE/TRAIT                 0\n",
      "INITIAL SAMPLE SIZE           0\n",
      "REPLICATION SAMPLE SIZE       0\n",
      "REGION                        0\n",
      "CHR_ID                        0\n",
      "CHR_POS                       0\n",
      "REPORTED GENE(S)              0\n",
      "MAPPED_GENE                   0\n",
      "UPSTREAM_GENE_ID              0\n",
      "DOWNSTREAM_GENE_ID            0\n",
      "SNP_GENE_IDS                  0\n",
      "UPSTREAM_GENE_DISTANCE        0\n",
      "DOWNSTREAM_GENE_DISTANCE      0\n",
      "STRONGEST SNP-RISK ALLELE     0\n",
      "SNPS                          0\n",
      "MERGED                        0\n",
      "SNP_ID_CURRENT                0\n",
      "CONTEXT                       0\n",
      "INTERGENIC                    0\n",
      "RISK ALLELE FREQUENCY         0\n",
      "P-VALUE                       0\n",
      "PVALUE_MLOG                   0\n",
      "P-VALUE (TEXT)                0\n",
      "OR or BETA                    0\n",
      "95% CI (TEXT)                 0\n",
      "PLATFORM [SNPS PASSING QC]    0\n",
      "CNV                           0\n",
      "MAPPED_TRAIT                  0\n",
      "MAPPED_TRAIT_URI              0\n",
      "STUDY ACCESSION               0\n",
      "GENOTYPING TECHNOLOGY         0\n",
      "Gene_Key                      0\n",
      "Entry                         0\n",
      "Reviewed                      0\n",
      "Entry Name                    0\n",
      "Protein names                 0\n",
      "Gene Names                    0\n",
      "Organism                      0\n",
      "Length                        0\n",
      "Involvement in disease        0\n",
      "Sequence_y                    0\n",
      "dtype: int64\n",
      "\n",
      "Cleaned Merged Data Info:\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 69728 entries, 0 to 69727\n",
      "Data columns (total 48 columns):\n",
      " #   Column                      Non-Null Count  Dtype  \n",
      "---  ------                      --------------  -----  \n",
      " 0   DATE ADDED TO CATALOG       69728 non-null  object \n",
      " 1   PUBMEDID                    69728 non-null  int64  \n",
      " 2   FIRST AUTHOR                69728 non-null  object \n",
      " 3   DATE                        69728 non-null  object \n",
      " 4   JOURNAL                     69728 non-null  object \n",
      " 5   LINK                        69728 non-null  object \n",
      " 6   STUDY                       69728 non-null  object \n",
      " 7   DISEASE/TRAIT               69728 non-null  object \n",
      " 8   INITIAL SAMPLE SIZE         69728 non-null  object \n",
      " 9   REPLICATION SAMPLE SIZE     69728 non-null  object \n",
      " 10  REGION                      69728 non-null  object \n",
      " 11  CHR_ID                      69728 non-null  object \n",
      " 12  CHR_POS                     69728 non-null  object \n",
      " 13  REPORTED GENE(S)            69728 non-null  object \n",
      " 14  MAPPED_GENE                 69728 non-null  object \n",
      " 15  UPSTREAM_GENE_ID            69728 non-null  object \n",
      " 16  DOWNSTREAM_GENE_ID          69728 non-null  object \n",
      " 17  SNP_GENE_IDS                69728 non-null  object \n",
      " 18  UPSTREAM_GENE_DISTANCE      69728 non-null  float64\n",
      " 19  DOWNSTREAM_GENE_DISTANCE    69728 non-null  float64\n",
      " 20  STRONGEST SNP-RISK ALLELE   69728 non-null  object \n",
      " 21  SNPS                        69728 non-null  object \n",
      " 22  MERGED                      69728 non-null  int64  \n",
      " 23  SNP_ID_CURRENT              69728 non-null  float64\n",
      " 24  CONTEXT                     69728 non-null  object \n",
      " 25  INTERGENIC                  69728 non-null  float64\n",
      " 26  RISK ALLELE FREQUENCY       69728 non-null  object \n",
      " 27  P-VALUE                     69728 non-null  float64\n",
      " 28  PVALUE_MLOG                 69728 non-null  float64\n",
      " 29  P-VALUE (TEXT)              69728 non-null  object \n",
      " 30  OR or BETA                  69728 non-null  float64\n",
      " 31  95% CI (TEXT)               69728 non-null  object \n",
      " 32  PLATFORM [SNPS PASSING QC]  69728 non-null  object \n",
      " 33  CNV                         69728 non-null  object \n",
      " 34  MAPPED_TRAIT                69728 non-null  object \n",
      " 35  MAPPED_TRAIT_URI            69728 non-null  object \n",
      " 36  STUDY ACCESSION             69728 non-null  object \n",
      " 37  GENOTYPING TECHNOLOGY       69728 non-null  object \n",
      " 38  Gene_Key                    69728 non-null  object \n",
      " 39  Entry                       69728 non-null  object \n",
      " 40  Reviewed                    69728 non-null  object \n",
      " 41  Entry Name                  69728 non-null  object \n",
      " 42  Protein names               69728 non-null  object \n",
      " 43  Gene Names                  69728 non-null  object \n",
      " 44  Organism                    69728 non-null  object \n",
      " 45  Length                      69728 non-null  float64\n",
      " 46  Involvement in disease      69728 non-null  object \n",
      " 47  Sequence_y                  69728 non-null  object \n",
      "dtypes: float64(8), int64(2), object(38)\n",
      "memory usage: 25.5+ MB\n",
      "None\n",
      "\n",
      "Numerical Data Summary:\n",
      "           PUBMEDID  UPSTREAM_GENE_DISTANCE  DOWNSTREAM_GENE_DISTANCE  \\\n",
      "count  6.972800e+04            6.972800e+04              6.972800e+04   \n",
      "mean   3.294042e+07            2.984523e+04              3.154085e+04   \n",
      "std    3.140909e+06            2.256574e+04              2.263643e+04   \n",
      "min    1.790329e+07            7.800000e+01              2.800000e+01   \n",
      "25%    3.321351e+07            2.867500e+04              3.053400e+04   \n",
      "50%    3.321351e+07            2.867500e+04              3.053400e+04   \n",
      "75%    3.538383e+07            2.867500e+04              3.053400e+04   \n",
      "max    3.790179e+07            1.205636e+06              1.422260e+06   \n",
      "\n",
      "             MERGED  SNP_ID_CURRENT    INTERGENIC       P-VALUE   PVALUE_MLOG  \\\n",
      "count  69728.000000    6.972800e+04  69728.000000  6.972800e+04  69728.000000   \n",
      "mean       0.000158    1.379647e+07      0.911814  4.005561e-06      6.495134   \n",
      "std        0.012559    2.081479e+07      0.283567  3.197839e-06      7.569770   \n",
      "min        0.000000    4.311000e+03      0.000000  0.000000e+00      5.000000   \n",
      "25%        0.000000    1.083817e+07      1.000000  9.000000e-07      5.154902   \n",
      "50%        0.000000    1.083817e+07      1.000000  4.000000e-06      5.397940   \n",
      "75%        0.000000    1.083817e+07      1.000000  7.000000e-06      6.045757   \n",
      "max        1.000000    1.795325e+09      1.000000  1.000000e-05    379.000000   \n",
      "\n",
      "         OR or BETA        Length  \n",
      "count  69728.000000  69728.000000  \n",
      "mean      38.217445    464.055559  \n",
      "std      595.718645    278.117514  \n",
      "min        0.004992     -1.000000  \n",
      "25%        0.104500    282.000000  \n",
      "50%        0.189100    444.000000  \n",
      "75%        0.416400    660.000000  \n",
      "max    15750.792000   2214.000000  \n",
      "\n",
      "Top 10 genes associated with Alzheimer's:\n",
      "Gene_Key\n",
      "Unknown           6026\n",
      "NECTIN2             90\n",
      "APOE                54\n",
      "BCAM - NECTIN2      46\n",
      "APP                 44\n",
      "CEACAM16-AS1        42\n",
      "TOMM40              42\n",
      "CLU                 40\n",
      "BIN1 - NIFKP9       32\n",
      "ABCA7               28\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Merged data saved to: C:\\Users\\14807\\OneDrive\\Spreadsheets\\Data Science Capstone\\AI_ML_Biotech\\Data Collection\\Final Data Merging after Collection\\merged_alzheimers_data.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\14807\\anaconda3\\Lib\\site-packages\\seaborn\\_oldcore.py:1119: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.\n",
      "  with pd.option_context('mode.use_inf_as_na', True):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete. Check the generated PNG files for visualizations.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Set the correct directory path\n",
    "data_dir = r\"C:\\Users\\14807\\OneDrive\\Spreadsheets\\Data Science Capstone\\AI_ML_Biotech\\Data Collection\\Final Data Merging after Collection\"\n",
    "\n",
    "# Load the protein data\n",
    "protein_data_path = os.path.join(data_dir, \"cleaned_protein_data.csv\")\n",
    "protein_data = pd.read_csv(protein_data_path)\n",
    "\n",
    "# Load the measurement data\n",
    "measurement_data_path = os.path.join(data_dir, \"cleaned_measurement_data.csv\")\n",
    "measurement_data = pd.read_csv(measurement_data_path)\n",
    "\n",
    "# Create Gene_Key for merging\n",
    "protein_data['Gene_Key'] = protein_data['Gene Names'].str.split().str[0]\n",
    "measurement_data['Gene_Key'] = measurement_data['MAPPED_GENE'].str.split(';').str[0]\n",
    "\n",
    "# Merge datasets\n",
    "merged_data = pd.merge(measurement_data, protein_data, on='Gene_Key', how='left')\n",
    "\n",
    "print(\"\\nMerged Data Info:\")\n",
    "print(merged_data.info())\n",
    "\n",
    "# Check for any columns that might have been duplicated during merge\n",
    "duplicated_columns = merged_data.columns[merged_data.columns.duplicated()]\n",
    "print(\"\\nDuplicated columns:\")\n",
    "print(duplicated_columns)\n",
    "\n",
    "# Remove duplicated columns\n",
    "merged_data = merged_data.loc[:, ~merged_data.columns.duplicated()]\n",
    "\n",
    "# Handle missing values\n",
    "print(\"\\nMissing values in each column:\")\n",
    "print(merged_data.isnull().sum())\n",
    "\n",
    "# For categorical columns\n",
    "categorical_columns = merged_data.select_dtypes(include=['object']).columns\n",
    "for col in categorical_columns:\n",
    "    if merged_data[col].isnull().sum() > 0:\n",
    "        if col in ['Gene_Key', 'MAPPED_GENE', 'REPORTED GENE(S)']:\n",
    "            merged_data[col] = merged_data[col].fillna('Unknown_Gene')\n",
    "        elif 'GENE' in col.upper():\n",
    "            merged_data[col] = merged_data[col].fillna('Unknown_Gene')\n",
    "        else:\n",
    "            merged_data[col] = merged_data[col].fillna('Unknown')\n",
    "\n",
    "# For numerical columns\n",
    "numerical_columns = merged_data.select_dtypes(include=['number']).columns\n",
    "for col in numerical_columns:\n",
    "    if merged_data[col].isnull().sum() > 0:\n",
    "        if col in ['P-VALUE', 'PVALUE_MLOG']:\n",
    "            merged_data[col] = merged_data[col].fillna(1)\n",
    "        elif col in ['OR or BETA']:\n",
    "            merged_data[col] = merged_data[col].fillna(1)\n",
    "        elif col in ['Length']:\n",
    "            merged_data[col] = merged_data[col].fillna(-1)  # Indicating unknown length\n",
    "        else:\n",
    "            merged_data[col] = merged_data[col].fillna(merged_data[col].median())\n",
    "\n",
    "# Check missing values again\n",
    "print(\"\\nMissing values after filling:\")\n",
    "print(merged_data.isnull().sum())\n",
    "\n",
    "print(\"\\nCleaned Merged Data Info:\")\n",
    "print(merged_data.info())\n",
    "\n",
    "# Basic statistics for numerical columns\n",
    "print(\"\\nNumerical Data Summary:\")\n",
    "print(merged_data.describe())\n",
    "\n",
    "# Top Alzheimer's related genes\n",
    "alzheimers_data = merged_data[merged_data['DISEASE/TRAIT'].str.contains('Alzheimer', case=False, na=False)]\n",
    "print(\"\\nTop 10 genes associated with Alzheimer's:\")\n",
    "print(alzheimers_data['Gene_Key'].value_counts().head(10))\n",
    "\n",
    "# Save the merged dataset\n",
    "output_path = os.path.join(data_dir, \"merged_alzheimers_data.csv\")\n",
    "merged_data.to_csv(output_path, index=False)\n",
    "print(f\"\\nMerged data saved to: {output_path}\")\n",
    "\n",
    "# Additional analysis and visualizations\n",
    "\n",
    "# 1. Distribution of p-values for Alzheimer's-related entries\n",
    "plt.figure(figsize=(10, 6))\n",
    "sns.histplot(alzheimers_data['P-VALUE'], bins=50, kde=True)\n",
    "plt.title(\"Distribution of P-values for Alzheimer's-related Entries\")\n",
    "plt.xlabel(\"P-value\")\n",
    "plt.savefig(os.path.join(data_dir, \"alzheimers_pvalue_distribution.png\"))\n",
    "plt.close()\n",
    "\n",
    "# 2. Protein lengths of top Alzheimer's-associated genes\n",
    "top_genes = alzheimers_data['Gene_Key'].value_counts().nlargest(10).index\n",
    "plt.figure(figsize=(12, 6))\n",
    "sns.boxplot(x='Gene_Key', y='Length', data=alzheimers_data[alzheimers_data['Gene_Key'].isin(top_genes)])\n",
    "plt.title(\"Protein Lengths of Top Alzheimer's-associated Genes\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.savefig(os.path.join(data_dir, \"alzheimers_protein_lengths.png\"))\n",
    "plt.close()\n",
    "\n",
    "# 3. Correlation heatmap of key numerical features\n",
    "correlation_columns = ['UPSTREAM_GENE_DISTANCE', 'DOWNSTREAM_GENE_DISTANCE', 'P-VALUE', 'PVALUE_MLOG', 'OR or BETA', 'Length']\n",
    "correlation_matrix = merged_data[correlation_columns].corr()\n",
    "plt.figure(figsize=(10, 8))\n",
    "sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm')\n",
    "plt.title(\"Correlation Matrix of Key Numerical Features\")\n",
    "plt.savefig(os.path.join(data_dir, \"correlation_matrix.png\"))\n",
    "plt.close()\n",
    "\n",
    "print(\"Analysis complete. Check the generated PNG files for visualizations.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b7e9799",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "NaN values after replacing inf:\n",
      "DATE ADDED TO CATALOG         0\n",
      "PUBMEDID                      0\n",
      "FIRST AUTHOR                  0\n",
      "DATE                          0\n",
      "JOURNAL                       0\n",
      "LINK                          0\n",
      "STUDY                         0\n",
      "DISEASE/TRAIT                 0\n",
      "INITIAL SAMPLE SIZE           0\n",
      "REPLICATION SAMPLE SIZE       0\n",
      "REGION                        0\n",
      "CHR_ID                        0\n",
      "CHR_POS                       0\n",
      "REPORTED GENE(S)              0\n",
      "MAPPED_GENE                   0\n",
      "UPSTREAM_GENE_ID              0\n",
      "DOWNSTREAM_GENE_ID            0\n",
      "SNP_GENE_IDS                  0\n",
      "UPSTREAM_GENE_DISTANCE        0\n",
      "DOWNSTREAM_GENE_DISTANCE      0\n",
      "STRONGEST SNP-RISK ALLELE     0\n",
      "SNPS                          0\n",
      "MERGED                        0\n",
      "SNP_ID_CURRENT                0\n",
      "CONTEXT                       0\n",
      "INTERGENIC                    0\n",
      "RISK ALLELE FREQUENCY         0\n",
      "P-VALUE                       0\n",
      "PVALUE_MLOG                   0\n",
      "P-VALUE (TEXT)                0\n",
      "OR or BETA                    0\n",
      "95% CI (TEXT)                 0\n",
      "PLATFORM [SNPS PASSING QC]    0\n",
      "CNV                           0\n",
      "MAPPED_TRAIT                  0\n",
      "MAPPED_TRAIT_URI              0\n",
      "STUDY ACCESSION               0\n",
      "GENOTYPING TECHNOLOGY         0\n",
      "Gene_Key                      0\n",
      "Entry                         0\n",
      "Reviewed                      0\n",
      "Entry Name                    0\n",
      "Protein names                 0\n",
      "Gene Names                    0\n",
      "Organism                      0\n",
      "Length                        0\n",
      "Involvement in disease        0\n",
      "Sequence_y                    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# ... (your existing code for data loading and merging)\n",
    "\n",
    "# After cleaning and before analysis/plotting\n",
    "import numpy as np\n",
    "\n",
    "# Replace inf values with NaN\n",
    "merged_data = merged_data.replace([np.inf, -np.inf], np.nan)\n",
    "\n",
    "# Check for any new NaN values\n",
    "print(\"\\nNaN values after replacing inf:\")\n",
    "print(merged_data.isnull().sum())\n",
    "\n",
    "# Handle any new NaN values if necessary\n",
    "# For example:\n",
    "# merged_data = merged_data.dropna()  # or\n",
    "# merged_data = merged_data.fillna(method='ffill')\n",
    "\n",
    "# Continue with your analysis and plotting\n",
    "# ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "118872ce",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
